DS-3940 SHERPA v2 API (main port)

kshepherd opened the pull request:
Developed by The Library Code with the support of Technische Universität Hamburg TUHH.

## References
* JIRA link: [DS-3940](https://jira.lyrasis.org/browse/DS-3940)
* SHERPA API key registration: https://v2.sherpa.ac.uk/romeo/api.html
* DSpace 6 PR: #2739 
* The [REST Contract](https://github.com/DSpace/Rest7Contract) entry for this is [External Authority Services](https://github.com/DSpace/Rest7Contract/blob/master/external-authority-sources.md), and is unchanged (though, neither SHERPA source is used as an example in this doc -- it could be added if requested)
* No Angular changes are made as part of this work, it is a refactor of external sources

## Description
This is a master port of #2739 to introduce new SHERPA v2 API data models and integrations, and to refactor existing work in external source providers (for journal and publisher lookups) to make use of these new classes. No Angular changes or REST changes are made. 
It is not an exact port of 6.x as the XMLUI and JSPUI changes are irrelevant, tests have been added, and the default API key is not commented out in dspace.cfg but is instead left blank (as it is required as a placeholder by spring bean configuration). Everything else should be the same (the DSpace 6 version was updated to use DSpace 7 code style rules).

## Instructions for Reviewers
List of changes in this PR:
* Removed old SHERPA model and integration classes and replaced with classes to work with [SHERPA v2 API](https://v2.sherpa.ac.uk/api/object-retrieval.html) 
* Refactored and renamed SherpaJournalDataProvider and SherpaPublisherDataProvider provider classes, to work with the new [SHERPA v2 API](https://v2.sherpa.ac.uk/api/object-retrieval.html) 
* Wrote two unit tests (one for Publisher, one for Journal) to ensure that JSON is parsed correctly. The JSON is taken from raw API responses and supplied as static files in the tests resources bundle

To test this PR:
* [Register for a SHERPA API v2 key](https://v2.sherpa.ac.uk/romeo/api.html) and configure it in local.cfg: **sherpa.romeo.apikey**
* Visit the external sources page in the rest HAL explorer: _http://${server.url}/server/api/integration/externalsources_
* The external sources **sherpaJournal** and **sherpaPublisher** should be listed (as they were previously with the v1 API code)
* For the sherpaJournal source, a search can be tested with a query for a journal name, eg. "PLOS" should return ~7 results.
* For the sherpaPublisher source, a search can be tested with a query for a publisher name, eg. "New Zealand" should return ~15 results
* The "get external object" query (exact title match) can be tested with eg. _http://${server.url}/server/api/integration/externalsources/sherpaJournal/entryValues/PLoS%20Computational%20Biology_

## Checklist
- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
(***My PR is longer, at around ~2000 including comments so probably ~1800-1900 without, but a lot of these changes seem longer since the refactor involves lots of files being replaced / renamed. The changes are isolated just to SHERPA integration and do not impact any other part of the DSpace codebase, so I am hoping this work can still be approved for review***)

- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide)
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types, including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for known error scenarios and error codes (e.g. `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`, etc)
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
(*It does not include any new dependencies*)
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.
(*It does not modify the REST API, but I have linked to the relevant contract page*)

kshepherd committed:
[DS-3940](master) Port base SHERPA v2 changes from 6.x version

kshepherd committed:
[DS-3940](master) Refactor external providers for journals, publishers

kshepherd committed:
[DS-3940](master) Unit tests for SHERPAResponse, SHERPAPublisherResponse

kshepherd committed:
[DS-3940](master) Configuration changes for new API key, URL, external svcs

kshepherd committed:
[DS-3940](master) Removal of legacy v1 SHERPA classes and data providers

kshepherd committed:
[DS-3940](master) Better handling of empty / null results in providers

kshepherd committed:
[DS-3940](master) Ensure sherpa.romeo.apikey placeholder always exists

Spring bean configuration (external-sources.xml) requires the placeholder,
even if it is null, so this change reverts the commented out api key
introduced with other v2 refactor work

kshepherd committed:
[DS-3940](master) Add license headers to unit test classes

kshepherd committed:
[DS-3940](master) Fixes for Checkstyle failures in unit tests
(imports, whitespace after catch)

abollini reviewed (commented):
Thanks @kshepherd for this PR, I have made a first review by code inspection and only have few suggestion inline. I would like to make a full test in the next days.
I would really appreciate the addition of an integration test because this would be the only way to assure a proper end to end implementation and also give us an indicator in future for issues related to changes on the SHERPA side.
I guess that right now some edge scenario where we got an invalid or error response from SHERPA are hidden and not reported correctly on the rest api (see comments about catch of exception without re-throw) and other issue, like missing escape of query parameters are not tested.
In DSpace-CRIS to test integration with external system that requires API KEY we use this trick (the key is stored as a secret env variable on travis) 
https://github.com/4Science/DSpace/blob/dspace-cris-7/.travis.yml#L25

KevinVdV commented on dspace-api/src/main/java/org/dspace/app/sherpa/submit/SHERPASubmitService.java:
```diff
@@ -37,12 +38,28 @@ public void setSherpaService(SHERPAService sherpaService) {
+                responses.add(response);
+            }
+            if (responses.isEmpty()) {
+                responses.add(new SHERPAResponse("SHERPA ROMeO lookup failed"));
```
Did the lookup "fail", it could be that there are no responses ?

kshepherd replied:
Yes - in this case "responses" is a linked list of each individual API response objects, not the list of items included in the JSON response. (because we're chaining some queries for multiple ISSNs). 
Even an "empty" API response should still be represented in a valid SHERPAResponse object.

KevinVdV replied:
You are correct, resolving.

KevinVdV commented on dspace-api/src/main/java/org/dspace/app/sherpa/submit/SHERPASubmitService.java:
```diff
@@ -37,12 +38,28 @@ public void setSherpaService(SHERPAService sherpaService) {
+                SHERPAResponse response = sherpaService.searchByJournalISSN(issn);
+                if (response.isError()) {
+                    // Continue with loop
+                    log.warn("Failed to look up SHERPA ROMeO result for ISSN: " + issn);
```
Would it be possible to log the response body as well ? (Not sure if this possible)

kshepherd replied:
It's not easily possible here, but what I've done is logged the error message (created upstream when error was set to true in the SHERPAResponse object), and also logged the response body in the SHERPAService and Data Providers if they ever encounter a response code that isn't HttpStatus.SC_OK
See commit: fdc5a3f2

KevinVdV replied:
Ok

KevinVdV commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2PublisherDataProvider.java:
```diff
@@ -0,0 +1,361 @@
+            get = constructHttpGet(query);
+            HttpClient hc = new DefaultHttpClient();
+            HttpResponse response = hc.execute(get);
+            if (response.getStatusLine().getStatusCode() == 200) {
```
Can we use "HttpStatus.SC_OK" instead of 200 ?

kshepherd replied:
Done! See commit# 839159b

KevinVdV commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2JournalDataProvider.java:
```diff
@@ -0,0 +1,358 @@
+            get = constructHttpGet(query, start, limit);
+            HttpClient hc = new DefaultHttpClient();
+            HttpResponse response = hc.execute(get);
+            if (response.getStatusLine().getStatusCode() == 200) {
```
Can we use "HttpStatus.SC_OK" instead of 200 ?

kshepherd replied:
Done! See commit# 839159b

KevinVdV commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2JournalDataProvider.java:
```diff
@@ -0,0 +1,358 @@
+            get = constructHttpGet(query);
+            HttpClient hc = new DefaultHttpClient();
+            HttpResponse response = hc.execute(get);
+            if (response.getStatusLine().getStatusCode() == 200) {
```
Can we use "HttpStatus.SC_OK" instead of 200 ?

kshepherd replied:
Done! See commit# 839159b:
https://github.com/DSpace/DSpace/pull/2773/commits/839159bc8511dff8e0b8b58eb24b60957a5bd7c0

KevinVdV commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2PublisherDataProvider.java:
```diff
@@ -0,0 +1,361 @@
+            get = constructHttpGet(query, start, limit);
+            HttpClient hc = new DefaultHttpClient();
+            HttpResponse response = hc.execute(get);
+            if (response.getStatusLine().getStatusCode() == 200) {
```
Can we use "HttpStatus.SC_OK" instead of 200 ?

kshepherd replied:
Done! See commit# 839159b

KevinVdV reviewed (dismissed):
Aside from the things Andrea already found, I also found a few small code tweaks that are needed after my initial code review.

kshepherd commented:
Thanks @abollini and @KevinVdV , I'll work on this feedback and re-request a review when ready

kshepherd committed:
[DS-3940](main) Replace '200' with HttpStatus.SC_OK in data providers

kshepherd committed:
[DS-3940](main) Add basic javadoc for new SHERPA data classes

kshepherd commented:
> In DSpace-CRIS to test integration with external system that requires API KEY we use this trick (the key is stored as a secret env variable on travis)

Thanks @abollini - I am checking with SHERPA devs to make sure they're OK with usage of their production API / API keys for testing like this, and if there are any other issues we need to think of.

So that I understand what you mean above - this means DSpace Travis CI builds will be able to run the integration test with an API key but that key is not visible to non-Travis admins, and so a dev wanting to run local maven tests would have to include their own API key or they'll see a maven test failure?
Would this possibly cause confusiong for devs who don't have a key / use SHERPA, but still want to see 100% test passing in DSpace? It would make SHERPA API registration a prerequisite for running local maven tests?
Or is there a way to ignore this test except for Travis CI builds?

KevinVdV commented:
@kshepherd @abollini Just my 2 cents on the integration tests: I think we 2 types need integration tests, one where we query the actual API & another where we mock the API responses.

The integration tests where we query the API should not run by default (as we don't want DSpace tests to crash if the API is unavailable). The use of these tests would be to check if the parsing still works with the latest responses from the API (if SHERPA makes updates there we will want to test DSpace as well).

tdonohue commented:
@kshepherd @abollini : I agree with @KevinVdV here.   Any integration tests which hit a "live" API _should **never** run by default in Travis CI or anywhere else._ They are only there as a resource to help update our API to align with any future Sherpa API changes, and honestly I see them as lower priority than the other type of integration test (see below).

The higher priority integration test (in my mind) is one which *always runs* against either a mock (or a static export file) of the Sherpa v2 API.  This just ensures our code doesn't break against what we currently know about the Sherpa API. It may not be as perfect as the integration test against the "live API", but it's more stable for Travis (as we wouldn't want strange Travis failures anytime Sherpa does maintenance on their API and/or it simply goes down temporarily).

kshepherd committed:
[DS-3940](main) Shift spring config to DSpace config service for providers

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2PublisherDataProvider.java:
```diff
@@ -0,0 +1,361 @@
+     */
+    public void setApiKey(String apiKey) {
+        this.apiKey = apiKey;
+    }
```
Done! See commit 6ce9283e
I also moved URL config for consistency

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2JournalDataProvider.java:
```diff
@@ -0,0 +1,358 @@
+     */
+    public void setApiKey(String apiKey) {
+        this.apiKey = apiKey;
+    }
```
Done! See commit 6ce9283e
I also moved URL config for consistency

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAJournal.java:
```diff
@@ -0,0 +1,99 @@
+
+import java.util.List;
+
+public class SHERPAJournal {
```
Done! See commit af615a37

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAPermittedVersion.java:
```diff
@@ -0,0 +1,102 @@
+package org.dspace.app.sherpa.v2;
+
+import java.util.List;
+
```
Done! See commit af615a3

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAPublisher.java:
```diff
@@ -0,0 +1,89 @@
+ */
+package org.dspace.app.sherpa.v2;
+
+public class SHERPAPublisher {
```
Done! See commit af615a3

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAPublisherPolicy.java:
```diff
@@ -0,0 +1,116 @@
+import java.util.List;
+import java.util.Map;
+
+public class SHERPAPublisherPolicy {
```
Done! See commit af615a3

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPASystemMetadata.java:
```diff
@@ -0,0 +1,69 @@
+ */
+package org.dspace.app.sherpa.v2;
+
+public class SHERPASystemMetadata {
```
Done! See commit af615a3

kshepherd committed:
[DS-3940](main) Sanitise query strings for SHERPA v2 API calls

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2PublisherDataProvider.java:
```diff
@@ -0,0 +1,361 @@
+        // Build URL based on search query
+        URIBuilder uriBuilder = new URIBuilder(url);
+        uriBuilder.addParameter("item-type", "publisher");
+        uriBuilder.addParameter("filter", "[[\"name\",\"contains word\",\"" + query + "\"]]");
```
After talking to SHERPA devs I have done some simple stripping for valid URI / JSON parsing. Note, the URI Builder itself already escapes/encodes for URL
See commit a81fe58

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2JournalDataProvider.java:
```diff
@@ -0,0 +1,358 @@
+            // Construct URI for an exact match on journal title
+            uriBuilder = new URIBuilder(url);
+            uriBuilder.addParameter("item-type", "publication");
+            uriBuilder.addParameter("filter", "[[\"title\",\"equals\",\"" + id + "\"]]");
```
After talking to SHERPA devs I have done some simple stripping for valid URI / JSON parsing. Note, the URI Builder itself already escapes/encodes for URL
See commit a81fe58

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2JournalDataProvider.java:
```diff
@@ -0,0 +1,358 @@
+        // Build URL based on search query
+        URIBuilder uriBuilder = new URIBuilder(url);
+        uriBuilder.addParameter("item-type", "publication");
+        uriBuilder.addParameter("filter", "[[\"title\",\"contains word\",\"" + query + "\"]]");
```
After talking to SHERPA devs I have done some simple stripping for valid URI / JSON parsing. Note, the URI Builder itself already escapes/encodes for URL
See commit a81fe58

kshepherd committed:
[DS-3940](main) Sanitise query string (bugfix for bad method name in Pub)

kshepherd committed:
[DS-3940](main) Rethrow IOExceptions that are caught within main try block

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2PublisherDataProvider.java:
```diff
@@ -0,0 +1,361 @@
+                        return sherpaResponse.getPublishers().size();
+                    }
+                } catch (IOException e) {
+                    log.error("Error reading input stream for SHERPAResponse: " + e.getMessage());
```
I have now rethrown this exception in 3ef47dac - this one I can do, as it is nested within a higher try/catch block.
However, note that for the main catch in the method:
This class and its methods implement ExternalDataService interface (via extending ExternalDataServiceImpl). The abstract methods don't have any throw statements in its signatures. I would be hesitant to include a refactor of this interface and abstract impl in this PR as it would have consequences for all other providers.

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2PublisherDataProvider.java:
```diff
@@ -0,0 +1,361 @@
+                        return list;
+                    }
+                } catch (IOException e) {
+                    log.error("Error parsing SHERPA response input stream: " + e.getMessage());
```
I have now rethrown this exception in 3ef47dac - this one I can do, as it is nested within a higher try/catch block.
However, note that for the main catch in the method:
This class and its methods implement ExternalDataService interface (via extending ExternalDataServiceImpl). The abstract methods don't have any throw statements in its signatures. I would be hesitant to include a refactor of this interface and abstract impl in this PR as it would have consequences for all other providers.

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2JournalDataProvider.java:
```diff
@@ -0,0 +1,358 @@
+                }
+            }
+        } catch (IOException | URISyntaxException e) {
+            log.error("SHERPA/RoMEO query failed: ", e);
```
I don't think we can.
This class and its methods implement ExternalDataService interface (via extending ExternalDataServiceImpl). The abstract methods don't have any throw statements in its signatures. I would be hesitant to include a refactor of this interface and abstract impl in this PR as it would have consequences for all other providers.

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2JournalDataProvider.java:
```diff
@@ -0,0 +1,358 @@
+                sherpaResponse = new SHERPAResponse("SHERPA/RoMEO returned no response");
+            }
+        } catch (Exception e) {
+            log.error("SHERPA/RoMEO query failed: ", e);
```
I don't think we can.
This class and its methods implement ExternalDataService interface (via extending ExternalDataServiceImpl). The abstract methods don't have any throw statements in its signatures. I would be hesitant to include a refactor of this interface and abstract impl in this PR as it would have consequences for all other providers.

kshepherd commented on dspace-api/src/main/java/org/dspace/external/provider/impl/SHERPAv2JournalDataProvider.java:
```diff
@@ -0,0 +1,358 @@
+                        return sherpaResponse.getJournals().size();
+                    }
+                } catch (IOException e) {
+                    log.error("Error reading input stream for SHERPAResponse: " + e.getMessage());
```
I have now rethrown this exception in 3ef47da - this one I can do, as it is nested within a higher try/catch block.
However, note that for the main catch in the method:
This class and its methods implement ExternalDataService interface (via extending ExternalDataServiceImpl). The abstract methods don't have any throw statements in its signatures. I would be hesitant to include a refactor of this interface and abstract impl in this PR as it would have consequences for all other providers.

kshepherd commented:
> @kshepherd @abollini : I agree with @KevinVdV here. Any integration tests which hit a "live" API _should **never** run by default in Travis CI or anywhere else._ 
This matches my thinking. If I have to write one, I'd prefer not to have it run with every build or by default. To make sure I write this correctly, it's just a case of extending **AbstractIntegrationTest** in a new **SHERPAServiceIT** test class?

kshepherd committed:
[DS-3940](main) Remove versionLabel from permittedVersion - just store key

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/v2/SHERPAResponse.java:
```diff
@@ -0,0 +1,533 @@
+            versionLabel = I18nUtil.getMessage("jsp.sherpa.accepted-version-label");
+        } else if ("published".equals(articleVersion)) {
+            versionLabel = I18nUtil.getMessage("jsp.sherpa.published-version-label");
+        }
```
This is a good point. Since the key ('submitted' or 'accepted' or 'published') is already stored in the articleVersion member variable and is available via a public getter as was the label, I have simply removed the versionLabel and its getters/setters to avoid any confusion.
The key can be used to build an i18n string.
See commit: 7cb1d1fa

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java:
```diff
@@ -91,12 +108,22 @@ public SHERPAResponse searchByJournalISSN(String query) {
                 }
             } catch (Exception e) {
-                log.warn("Encountered exception while contacting SHERPA/RoMEO: " + e.getMessage(), e);
+                log.error("Encountered exception while contacting SHERPA/RoMEO: " + e.getMessage(), e);
```
We have to handle at least 3 exception types here (Interrupted, for sleep, URISyntax, and IO) - and so I guess rethrow each type? Or is there a new exception type you think I should throw?
Unlike the provider classes, this one is OK to change in terms of exception usage since it isn't currently used by anything (presumably the REST API can use it in future), and doesn't implement any superclasses.
If you want me to rethrow each exception type in its own catch, that's fine. If you want me to use Exception, that's fine too. There must be one that works best though... HttpResponseException?

kshepherd committed:
[DS-3940](main) Log response body in HTTP errors, err msg in submit service

kshepherd commented on dspace-api/src/main/java/org/dspace/app/sherpa/SHERPAService.java:
```diff
@@ -91,12 +108,22 @@ public SHERPAResponse searchByJournalISSN(String query) {
+                        content = responseBody.getContent();
+                        sherpaResponse = new SHERPAResponse(content, SHERPAResponse.SHERPAFormat.JSON);
+                    } catch (IOException e) {
+                        log.error("Encountered exception while contacting SHERPA/RoMEO: " + e.getMessage(), e);
```
see earlier content - i can rethrow the internal exceptions but if you want the whole method to throw an exception(s) i just have some questions as to which is the best. the v1 version of SHERPAService did not rethrow any exceptions so i didn't implement anything new for this refactor.

tdonohue commented:
@kshepherd : The DSpace 7 team has started a "testing guide" to refer to when writing tests. https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide

But, yes, it should be just creating a new integration test overall.  However, for the "live" test, in order to ignore it by default, we may want to name it differently.  Any tests that end in "IT" will be run by default, so we might name it something like `SherpaServiceLiveTest`.  That ensures it won't execute by default, but we choose to run it manually by passing `-Dtest=[fullpackage].SherpaServiceLiveTest` on the commandline.

kshepherd committed:
[DS-3940](main) Remove version label, getter, setter from PermittedVersion

kshepherd commented:
hi @tdonohue @KevinVdV @abollini - I just wanted to follow up on this. I'm hoping that we can use the mock SHERPA API integration test (JSON taken from a real API response) for now.
I could open a separate JIRA issue to add a live HTTP (non-default) IT later, to plan and review whether (and how) we should implement that IT in a future PR?

My reasoning here is: 
1. As far as I know, no other DSpace 7 integration tests contact live APIs so my assumption is that this isn't a strict requirement.
2. It wouldn't be run by default anyway or in Travis because of the reasons Tim listed above, which in my opinion makes it a bit less useful for catching code / parsing issues and more likely to catch false positives with network errors, etc.
3. I think it's more important now to get a fixed SHERPA service (which does have a working IT, using the offline JSON) into DSpace 7 before we miss out on next release, or introduce too many conflicts

Related to #3, I can see already there's a conflict introduced since my PR - in this case I think an easy one, my PR deletes that particular class so it's probably just an LGTM fix that I can check and apply to the new class if relevant

Please let me know what you all think. Cheers!

mwoodiupui commented:
I would say that it is simply *not possible* to write a correct test using data or behaviors which are not under the control of the test.  That means:  no use of external services such as SHERPA.  It's quite reasonable, however, to build development tools to check whether our simulation of an external service is still correct.  But, as others have said, those tools are not normal tests and are not run in the normal test set.

I think that the purest notion of "integration testing" comes down to us from people who were building things like payroll programs or rocket guidance systems, where every bit is in-house.  We need a different model, and probably a different term, for the automated testing of complete code paths in a body of code which exists to interface with external resources.

kshepherd commented:
> I would say that it is simply _not possible_ to write a correct test using data or behaviors which are not under the control of the test. That means: no use of external services such as SHERPA. It's quite reasonable, however, to build development tools to check whether our simulation of an external service is still correct. But, as others have said, those tools are not normal tests and are not run in the normal test set.
> 
> I think that the purest notion of "integration testing" comes down to us from people who were building things like payroll programs or rocket guidance systems, where every bit is in-house. We need a different model, and probably a different term, for the automated testing of complete code paths in a body of code which exists to interface with external resources.

Thanks @mwoodiupui , I agree there

tdonohue commented:
@kshepherd : I agree with your direction & with what @mwoodiupui said.  I see no reason to delay this for a "live HTTP Integration Test".  I also feel strongly (as stated previously) that those sorts of ITs are of very limited use.  They let us know if the external service *changed* their API, but we'd need to disable it by default in Travis CI to avoid random build/failures (if that external API is down temporarily).

So, all in all, we only require (static based) Integration Tests for new code. There is no requirement to build anything more than that, unless the developer feels it necessary.

kshepherd commented:
Thanks all! to my reviewers @abollini and @KevinVdV - based on this feedback, and my changes made in response to your earlier comments and requests, do you approve of merging?

KevinVdV reviewed (commented):
@kshepherd The code is looking good, but a question on testing (maybe @tdonohue can pitch in here as well): I currently see tests to check that our parsing is working. This is good, but I would also like to see tests for the following:
* Test for the SHERPASubmitService where we provide an item with no issn
* Test for the SHERPASubmitService where we provide an item with a valid issn
* Test for the SHERPASubmitService where we provide an item with an invalid issn

This way we can test the behavior of this class, the matching from issn to a sherpa response can be mocked in the SHERPAService.

tdonohue commented:
@kshepherd : just a note...I'm rescheduling this for 7.0 beta 5 (beta 4 is almost out the door).  If you could add the tests that @KevinVdV mentions (see previous comment) then I think this should be almost ready to go.

tdonohue commented:
@kshepherd : Just a friendly reminder that it'd be good to get this PR into `main` branch in the 7.0 beta 5 release....that will be the final beta prior to 7.0, and we'll be doing the Testathon coinciding with beta 5.  So, if possible, would you be able to help move this forward in the next weeks or month?  (Beta 5 development has just begun)

kshepherd commented:
sure thing, i'll make another reminder

kshepherd commented:
@KevinVdV regarding mocking the valid response for SHERPASubmitService tests - are there any examples or patterns in other dspace ITs to see how this should be implemented?
Looking at what I've seen so far, it seems to me that I'd actually mock the SHERPAService, which makes the actual http calls, and has spring configuration? so the mock response would return... a valid but minimal SHERPAResponse obj for validation? a response based on parsing the static example JSON?

tdonohue commented:
@kshepherd : In case it's helpful, here's one example already in the codebase: the [`MockCCLicenseConnectorServiceImpl`](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/test/java/org/dspace/license/MockCCLicenseConnectorServiceImpl.java) which mocks the expected behavior of the CC License API by extending the [`CCLicenseConnectorServiceImpl`](https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/license/CCLicenseConnectorServiceImpl.java).   There's also a very basic [`MockOpenUrlServiceImpl`](https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/statistics/export/service/MockOpenUrlServiceImpl.java).

Honestly, the tests you already have here are great.  But, I think what @KevinVdV was asking about was simply mocking the service to verify *our code* works as expected even when ISSNs are missing or invalid, etc.  That could involve simply testing that our code builds a specific type of querystring to send to Sherpa, or maybe it does nothing (if that's expected), or perhaps log an expected error/warning.

kshepherd committed:
[DS-3940] Additional javadoc, integration and unit tests, service refactor

kshepherd commented:
@tdonohue while writing some tests i've ended up wanting to refactor a few things to abstract the code that builds the URI requests across the sherpa service and all providers... nearly done but it is just adding a bit of time and might mean another inspection is needed. hope i'm still in time for beta 5?

tdonohue commented:
@kshepherd : yes, you have time. Beta 5 is unlikely to be released until sometime in Dec at the earliest, and it may be early Jan (cause of time off around holidays). Nonetheless, I'd recommend a large PR like this get in as early as plausible....that gives us more time to resolve any issues that arise.  So, let us know when you are ready for a re-review! Thanks!

KevinVdV commented:
@kshepherd I have been so busy the last couple of days, so it took me a while to get here. What I would do test it is to create a mock class for the SHERPAv2JournalDataProvider which would use hardcoded xml files instead of contacting an external url. That way the result will be always the same & you can code your tests against that.

tdonohue commented:
@kshepherd : Reminder, this PR still needs to be finalized/updated for 7.0 Beta 5 (the final beta), which is likely to be released in February.  It'd be a shame to release this update in 6.4, only to _not support_ it in 7.0.  Any chance we could get this quickly updated in time for testing/release in Beta 5 & 7.0 Final?  Is there additional support you need, or do we need to find a new steward for this work?

kshepherd committed:
[DS-3940] Remove unused import in SHERPAv2JournalDataProvider

kshepherd committed:
[DS-3940] Update ExternalSourcesRestControllerIT to count for new external sources

kshepherd committed:
[DS-3940] Correct typo in ExternalSourcesRestControllerIT

kshepherd commented:
Hi @tdonohue @KevinVdV @abollini  - I've added tests (and more javadoc) and I've also refactored the Data Providers and the SHERPAService so that all the actual HTTP requests are done from the SHERPAService..

To help with testing, I've made a MockSHERPAService that reads in JSON for a journal or publisher and returns the parsed response object instead of querying the live SHERPA API

The data providers and the (legacy?) SHERPASubmitService are then configured in test spring config to use MockSHERPAService

I have the following test classes:

- SHERPAServiceTest will do thorough inspection and testing of the response objects to make sure all expected fields are present, and also test URI construction and the specific 'search by ISSN' method
- SHERPASubmitServiceTest will construct an item with an ISSN and make sure that the "searchRelatedJournals" method ultimately returns the expected Lancet data
- SHERPADataProviderTest will test the getExternalObject and searchExternalDataObject methods of the Journal, JournalISSN, and Publisher data providers. Since the providers are configured to use the MockSHERPAService, we know to expect Lancet (journal) or PLOS (publisher) data from the test resource JSON

As well as these tests, the data providers are also included in the ExternalSourceRestControllerIT integration test, to make sure their endpoints resolve and that they appear correctly in spring configuration

I've also ported in a bugfix that I've since made to my local codebase, to properly handle cases where a permitted_oa node can have multiple instances of article versions.

I rebased against latest main branch before force pushing - there were some conflicts but they were just small changes I'd already covered anyway (moving to ConfigurationService, LogManager, etc.)

Please let me know how things are looking to you.

abollini reviewed (approved):
thanks @kshepherd I have reviewed the code, the tests and make a quick manual test on the REST side. Everything seems ok, I'm +1 to merge

tdonohue reviewed (approved):
👍 Looks great now @kshepherd !  Thanks for your hard work in getting this ported with tests, etc.  I've verified that all of @KevinVdV 's prior requests seem to be addressed.  I've also checked the code & tests and I have no concerns.  Finally, I did a test from the UI and it seems to be working! (Sidenote: in today's DSpace 7 meeting I mentioned that I tried to sign up for a Sherpa API Key and didn't get one sent to me last week.  Today, I tried again and managed to get it to work...so I was able to finally test this!)

All in all, this looks good and is at +2.  So, I'll merge it immediately!
